src/report.py

```diff
@@ -71,37 +64,47 @@ def make_report(
     report_data = []
 
     for gl in gain_loss:
-        report_data.append((gl["name"], gl["shares"], prices[gl["name"]], gl["change"]))
+        report_data.append((gl.name, gl.shares, prices[gl.name], gl.change))
     return report_data
 
 
-def get_report(report_data: List[Tuple[str, int, float, float]]) -> List[str]:
+def get_report(
+    report_data: List[Tuple[str, int, float, float]],
+    formatter: TableFormatter,
+) -> List[str]:
     """
     Returns a report as a table.
 
     :param report_data: List of formatted portfolios
+    :param formatter: Formatter object to format the report.
     :return: Returns the table as a list where each entry is a row of the table
     """
-    output = []
-    output.append(f"{'Name':>10s} {'Shares':>10s} {'Price':>10s} {'Change':>10s}")
-    output.append(" ".join(["-" * 10] * 4))
-    for r in report_data:
-        output.append("{:>10s} {:>10d} {:>10.2f} {:>10.2f}".format(*r))
-    return output
+    return formatter.format(("Name", "Shares", "Price", "Change"), report_data)
 
 
 def portfolio_report(
-    portfolio_file_path: Union[str, Path], prices_file_path: Union[str, Path]
+    portfolio_file_path: Union[str, Path],
+    prices_file_path: Union[str, Path],
+    fmt: str = "txt",
 ) -> List[str]:
     """
     Create a tabulated report from a portfolio file and price file.
 
     :param portfolio_file_path: Path to the portfolio file
     :param prices_file_path: Path to the prices file
+    :param fmt: Format for the report to printed as. Can be one of 'txt' or 'csv'
     :return: Returns the table as a list where each entry is a row of the table
     """
+    formatter: Union[TextTableFormatter, CSVTableFormatter]
+    if fmt == "txt":
```

Comment:
Your code is working fine currently with this if / elif / else statement but we can improve it to handle automatically new formatter. It is overkill for this use case but it is a good exercise!

If you have to add a new formatter, currently you have to do:
- add the corresponding class in `tableformat.py`
- update the `portfolio_report` from `report.py` with a new `elif` to support this new formatter

In general it is better to have to update only one file to add a new functionality, in order to do this you can compute a mapping format -> formatter class at runtime and use it in `portfolio_report`. Thus, for a new formatter you will only add the corresponding class and then it will be handle automatically in the rest of the code.

Here what should be done:
1. since each formatter corresponds to a given format, the format can be viewed as classattribute of formatter classes
2. list automatically the formatter classes: you know that they all inherit from TableFormatter
3. define the mapping format -> formatter class

Regarding 1.: you simply have to add `format` (or `fmt` to match your naming) class attribute to the formatter class. For the base class you can define an abstract property, it will allow to raise an error if the new formatter does not have the class attribute
```python
class TableFormatter(ABC):
    @property
    @abstractmethod
    def fmt(self) -> str:
          pass

class TextTableFormatter(TableFormatter):
        fmt = "txt"

class CSVTableFormatter(TableFormatter):
        fmt = "csv"
```

Regarding 2. and 3. you can list all the objects defined / imported in a file thanks to [inspect.getmembers](https://docs.python.org/3/library/inspect.html#inspect.getmembers) and you can also provide a function to apply some filtering on the object to retrieve. 

In our case we want to list formatter classes except `TableFormatter` (since it is the base class and it is not usable) so the method to implement must check 3 things:
1. is it a class?
2. is it a subclass of `TableFormatter`?
3. isn't it abstract ? This step is necessary since `TableFormatter` is considered as a subclass of `TableFormatter`
```python
def is_formatter(x: object) -> bool:
     return inspect.isclass(x) and issubclass(x, TableFormatter) and not inspect.isabstract(x)
```

This method should be fed to `inspect.getmembers`. In addition to the method you should provide the module (i.e. the file to inspect) this can be done by using the built-in package [sys](https://docs.python.org/3/library/sys.html) and the variable `__name__` (which contains the name of the current file):
```python
inspect.getmembers(sys.modules[__name__], predicate=is_formatter)
```

The output of the line just above is a list of tuple (class_name_as_str, class) so in our case in will be:
```
[("CSVTableFormatter", CSVTableFormatter), ("TextTableFormatter", TextTableFormatter)]
```

Then we can create the mapping format -> formatter class by doing:
```python
FORMAT2FORMATTER_CLS = {
    formatter_cls.fmt: formatter_cls for _, formatter_cls in inspect.getmembers(sys.modules[__name__], predictate=is_formatter)
}
```
:warning: this variable should be defined at the bottom of the file `tableformat.py` (i.e. after the definition of all the formatter classes) and not at the top (i.e. before classes definition). Indeed, the `inspect.getmembers` only list what is defined before the call to this function!

Finally, you can import this variable `FORMAT2FORMATTER_CLS` into `report.py` and update the `portfolio_report` as follows:
```python
def portfolio_report(
    portfolio_file_path: Union[str, Path],
    prices_file_path: Union[str, Path],
    fmt: str = "txt",
) -> List[str]:
    try:
        formatter = FORMAT2FORMATTER_CLS[fmt]
    except KeyError:
         raise RuntimeError(f"Unknown format {fmt} - available formats are {FORMAT2FORMATTER_CLS.keys()}")
```

With this implementation, when you want to add a new formatter you simply have to create a new formatter class which inherits from `TableFormatter` with the required methods to implement (i.e. `headings` / `rows`) + the classattribute to set (i.e. `fmt`)

-----------

I hope everything is understandable and do not hesitate if you have any question!  Not sure you should implement this for your use case but at least you should keep in much such functionality

Reply:
This is very cool. I'll try implement it.

Reply:
cool, don't hesitate if you have any questions!

Reply:
I tried this and it worked really well. Just needed to change `formatter = FORMAT2FORMATTER_CLS[fmt]()`.